Update classifier serialization

k-dominik opened the pull request:
First step in moving away from pickle-based classifier deserialization: regex-based deserialization of the pickled strings for classifiers, and classifier factories. I've checked our code history in order to find out which members of these classes need deserializing and added regular expressions for all of them. The code is quite verbose but tests cover all of it).

Luckily, all pickles in ilastik are with protocol0: human readable - so it was reasonably easy to parse.

While it is a lot of code, the idea is pretty simple - deserialize the objects in two steps:

1.) get the necessary values (types, attributes) out of the pickled string
2.) create instances / types.

see #2844 for the complete plan

---

Other stuff that slipped in (could very well be an extra PR..., but I don't expect any discussion there):
* removal of an unused (never in ilastik) classifier: `VigraRfAdaptiveMaskPixelwiseClassifier`
* removal of `getOrCreateGroup`, as `h5py` does have equivalent functionality (`require_group`)
* felt uneasy about `slice` (de)serialization: added tests that at least properly document the current behavior and  also raising now for conditions that we would not be able to represent with those serializers.

---

Todos:

- [x] Format code and imports.
- [x] Add docstrings and comments.
- [x] Add tests.
- [x] Reference relevant issues and other pull requests.
- [x] Rebase commits into a logical sequence.
- n/a: Update [user documentation](https://github.com/ilastik/ilastik.github.io).

btbest commented on tests/test_ilastik/test_applets/base/test_serializerUtils.py:
```diff
@@ -0,0 +1,552 @@
+        _ = deserialize_legacy_classifier_type_info(ds)
+
+
+@pytest.mark.parametrize(
+    "classifier_type, c_args, c_kwargs, expected_info",
+    [
+        (
```
Apparently this parameter collection is duplicated in lines 244+, could extract them into a variable. Or merge the two tests. Or remove the tests for the private functions altogether since they are covered by the tests on the public functions...

Similarly for `test_deserialize_classifier_factory_details` and `test_deserialize_legacy_classifier_factory`.

k-dominik replied:
I looked at it with Fynn and we decided to keep the duplicated parameters.

btbest reviewed (changes requested):
Wow, doing what must be done for security here. Great to see! Also appreciate the better typing.

Regarding serializerUtils:

- Naming would be better if:
  - "legacy" appears only in the public function names; as far as I can gather it's not that some of these classifiers are legacy and others are not, but all functions relating to deserializing in this file are about legacy.
  - `*_details` is removed from function names. There are a bunch of functions that return `*Info` objects, and some of them have names ending in `_details`, while others don't have this suffix. Consistency would be better; I prefer without the suffix.
- The amount of deserialization code here would warrant a `deserialize_legacy` submodule, separate from `deleteIfPresent` and the other generic utils at the top of the file. This would even be a case where I would tend towards grouping related functions in a class purely for the sake of code folding/traversing. E.g. one for the two classes and functions related to Vigra random forests, and one for the classes and functions related to sklearn.

test_serializerUtils could also benefit from the same split between generic utils and deserialization, and/or subgrouping tests into classes. As mentioned in one of the sub-comments, I think a bunch of these tests are duplicate coverage and could be removed. As it is, I haven't looked through all the tests in detail, but will tomorrow.

k-dominik commented on ilastik/applets/base/appletSerializer/serializerUtils.py:
```diff
@@ -18,31 +18,59 @@
+    if any(sl.step not in [None, 1] for sl in slicing):
+        raise ValueError("Only slices with step size of `1` or `None` are supported.")
+
+    if any(sl.start == None for sl in slicing):
+        raise ValueError("Start indices for slicing must be integer, got `None`.")
+
+    if any(sl.stop == None for sl in slicing):
```
nope! thx :)

k-dominik commented on ilastik/applets/base/appletSerializer/serializerUtils.py:
```diff
@@ -55,21 +83,574 @@ def slicingToString(slicing):
+    return ds[()].decode()
+
+
+LazyflowClassifierABCs = Union[LazyflowPixelwiseClassifierABC, LazyflowVectorwiseClassifierABC]
+
+LazyflowClassifierTypeABCs = Union[Type[LazyflowPixelwiseClassifierABC], Type[LazyflowVectorwiseClassifierABC]]
```
I don't feel strongly about this. `Union` is a set like operation (to me), and while sets can hold also only one element, usually they have multiple elements, as in here. I see it a bit like binary flag_s_, not that this is the same, but having something like `0010 | 0011`, would this for you constitute a single flag, or multiple ones?

On the other hand for a type, where one would only expect a single instance/type for an argument, it's probably better to name it so. `typing` does `Iterator` and not `Iterators` for example. So yeah, I guess it's best to change. Haha, now I do feel strongly about it.

k-dominik commented on ilastik/applets/base/appletSerializer/serializerUtils.py:
```diff
@@ -55,21 +83,574 @@ def slicingToString(slicing):
+        \.                                 # all pickles end in "." STOP
+        $
+    """,
+        re.X,
```
It's the only way (imo).

k-dominik commented on ilastik/applets/base/appletSerializer/slotSerializer.py:
```diff
@@ -828,10 +842,8 @@ def shouldSerialize(self, group):
         try:
-            # Attempt to unpickle
-            value = pickle.loads(pickled)
+            value = deserialize_legacy_classifier_factory(dset)
```
Nice find! The classifier factory would be crucial when retraining, so this _is_ needed. Will add a test!

k-dominik commented on ilastik/applets/base/appletSerializer/serializerUtils.py:
```diff
@@ -55,21 +83,574 @@ def slicingToString(slicing):
+
+@dataclass
+class ClassifierInfo:
+    submodule_name: str
+    type_name: str
```
Bummer, that this doesn't help with typing, I would prefer your proposed option if it worked.

btbest replied:
The way it's written above does work, and I like it. I'm not sure if there's a lot of benefit, besides a bit more traceability of what strings are which, but if you also like it then why not :D

k-dominik replied:
it doesn't work for type checking, which would be priority for me. It does work at runtime, yes.

k-dominik commented:
Thanks a lot @btbest for the sharp eye! Great and useful suggestions, also with regards to splitting the module up (will do). Will give tests a second look, but in general if any functionality is added by one of the wrapping functions I'd prefer to keep the tests, even with duplicate parameter lists.

k-dominik commented on ilastik/applets/base/appletSerializer/slotSerializer.py:
```diff
@@ -86,23 +100,23 @@ def __init__(self, slot, inslot=None, name=None, subname=None, default=None, dep
         self.subname = subname
 
-        self._dirty = False
+        self._dirty: bool = False
```
yeah these are redundant for the type checker, but i don't think it matters.

k-dominik commented on ilastik/applets/base/appletSerializer/serializerUtils.py:
```diff
@@ -55,21 +83,574 @@ def slicingToString(slicing):
+    if pickle_string.isascii() and (m := classifier_factory_pickle_string_matcher.search(pickle_string)):
+        groupdict = m.groupdict()
+        submodule = groupdict["factory_submodule"]
+        typename = groupdict["type_name"]
```
wow, didn't know that :)

k-dominik commented on ilastik/applets/base/appletSerializer/serializerUtils.py:
```diff
@@ -18,31 +18,59 @@
+
     The result is a utf-8 encoded bytes, for easy storage via h5py
     """
+    if any(sl.step not in [None, 1] for sl in slicing):
```
+1 on the converting early - but for now, this is out of scope for this PR.

k-dominik commented on ilastik/applets/base/appletSerializer/serializerUtils.py:
```diff
@@ -18,31 +18,59 @@
+    if any(sl.step not in [None, 1] for sl in slicing):
+        raise ValueError("Only slices with step size of `1` or `None` are supported.")
+
+    if any(sl.start == None for sl in slicing):
```
I replicated the behavior that was there in this functionality before. I don't think I'd want to broaden support now to other kinds of objects preemptively, without knowing if these ever occur in ilastik.

k-dominik commented:
Hello @emilmelnikov,

thank you very much for looking into this. It felt good having you as a co-author for a commit again :)

With respect to the slicing changes, they should've not been part of this PR I think. I really didn't want to improve those. I added test to document the current behavior, which has been in the field for many years now. Improvements are of course nice, yeah, just too much out of scope for now.

wrt to the pickling. It's really cool that you found this solution. If we were to write something generic I would want to jump on it immediately. The thing is that we want to move away from pickle in general for serialization - so this PR adds functionality to load exactly only that data, that we expect. It was deliberate not to allow any kind of object. Some other concerns would be using a private, not well documented class `pickle._Unpickler` that can change  with any release. And again, it just does so much more than we want.

k-dominik commented:
K, first of all thanks again @btbest, @emilmelnikov and @FynnBe for the additional pair of eyes and discussions. A quick summary of the major post-review changes:

* moved the legacy deserialization to the designated module `ilastik.applets.base.appletSerializer.legacyClassifiers`
* added a test for the actual classifier factory deserializer
* streamlined function naming
* direct access for groups on `re.match` objects

btbest commented:
The code changes look fine as far as I can tell, but I ran into a problem during manual testing. I created a pixel-classification project with a K-Neighbors classifier, trained, saved, closed, and then got this error message trying to reopen the project:

```
    super(SerialClassifierSlot, self).deserialize(group)
  File "\$path\ilastik\applets\base\appletSerializer\slotSerializer.py", line 218, in deserialize
    self._deserialize(group[self.name], self.inslot)
  File "\$path\ilastik\applets\base\appletSerializer\slotSerializer.py", line 649, in _deserialize
    classifier_type = deserialize_classifier_type(classifierGroup["pickled_type"])
  File "\$path\ilastik\applets\base\appletSerializer\legacyClassifiers.py", line 106, in deserialize_classifier_type
    class_string: str = deserialize_string_from_h5(ds)
  File "\$path\ilastik\applets\base\appletSerializer\serializerUtils.py", line 89, in deserialize_string_from_h5
    return ds[()].decode()
AttributeError: 'numpy.void' object has no attribute 'decode'
```

I guess either the K-Neighbors classifier doesn't get serialized correctly, or the regex in the new deserialization doesn't match.

btbest reviewed (approved):
Works now. Lucky that I happened to try exactly a kind of classifier that ends up being saved differently...
